### PR TITLE
[CELEBORN-1398] Support return leader ip to client

### PR DESCRIPTION
**Pull request by RexXiong**
### What changes were proposed in this pull request?
As title

### Why are the changes needed?
Currently, if accessing services of a Celeborn cluster across Kubernetes clusters, one may encounter DNS resolution issues. However, connectivity may be achieved through IP addresses when combined with the Kubernetes setting hostNetwork=true for clients from different clusters. At present, the `celeborn.network.bind.preferIpAddress` configuration is only effective on worker nodes. This PR will enable the feature of returning the leader's IP when accessing the master node.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GA

### CODE REVIEW

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/HARaftServer.java [hunk 1]
```diff
@@ -492,11 +494,19 @@ public void updateServerRole() {
     }
   }
 
+  public Pair<String, String> addressToIpHostAddressPair(String address) {
```
Comment (pan3793) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/HARaftServer.java:497]: a little bit confusing, can u add java docs and test for this method?

File: common/src/main/java/org/apache/celeborn/common/client/MasterNotLeaderException.java [hunk 2]
```diff
@@ -22,6 +22,7 @@
 import javax.annotation.Nullable;
 
 import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
 
 public class MasterNotLeaderException extends IOException {
```
Comment (pan3793) [inline common/src/main/java/org/apache/celeborn/common/client/MasterNotLeaderException.java:27]: the change in this class should be safe, but better to have a manual test to ensure that 0.4 client works well with main branch service

File: common/src/main/scala/org/apache/celeborn/common/util/Utils.scala [hunk 3]
```diff
@@ -432,6 +432,29 @@ object Utils extends Logging {
...
+    }
+  }
+
+  // Convert address (ip:port or host:port) to (ip:port, host:port) pair
+  def addressToIpHostAddressPair(address: String): (String, String) = {
+    val hostPort = Utils.parseHostPort(address)
+    val internalIpHostAddressPair = Utils.getIpHostAddressPair(hostPort._1)
+    (
+      internalIpHostAddressPair._1 + ":" + hostPort._2,
+      internalIpHostAddressPair._2 + ":" + hostPort._2)
```
[the remark this answers is not in the document]
Comment (RexXiong) [reply common/src/main/scala/org/apache/celeborn/common/util/Utils.scala]: Thanks, done.

File: common/src/main/java/org/apache/celeborn/common/client/MasterNotLeaderException.java (hunk 2 above)
Comment (RexXiong) [reply common/src/main/java/org/apache/celeborn/common/client/MasterNotLeaderException.java:28]: Yes, it's compatible with 0.4~

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/HARaftServer.java (hunk 1 above)
Comment (RexXiong) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/HARaftServer.java]: Move to Utils and add doc for this. Seems impossible mock object for Scala，so I do manual test for both Scenario.

**Review by pan3793 — APPROVED** [review]

**Review by SteNicholas — APPROVED** [review]

**Comment by RexXiong** [comment]
Thanks. merge to main(v0.5.0) & branch-0.4(v0.4.1)

**Pull request closed without merging**